### PR TITLE
Support macros in static Gen functions and within `@trace` calls.

### PR DESCRIPTION
**Pull request by ztangent**
This PR addresses #185 by supporting macros in the static modeling language, supersedes #278 by supporting macros nested within `@trace` calls in both the dynamic and static modeling languages, and supersedes the implementation strategy used in #223 to ensure correct scoping for `x ~ gen_fn()` tilde expressions.

More broadly, the changes made by this PR ensure that macro-handling is as close to Julia's semantics as possible, by making Julia handle all macro-expansion in the function body AST. This ensures that any issues with macros in generative function definitions can be traced to issues with the macro itself (or with Julia), and not to issues with Gen's parsing code. This should make future extensions to the modeling languages easier to implement and reason about.

At a high-level, the implementation strategy here enforces a separation between the surface syntax associated with Gen-specific macros (i.e., `@trace` and `@param`) and their corresponding implementations, which differ across the SML (`StaticIR` nodes) and DML (`traceat`, `splice`, or `read_param` calls). It does this by introducing custom expressions of the form `Expr(:gentrace, call, addr)` and `Expr(:genparam, name, type)`, which serve as intermediate representations in the macro-expanded AST.

Each modeling language can then handle these custom expressions in their own manner, either by parsing them to DAG nodes (for the SML) or by substituting them with their implementations (for the DML). This effectively allows the SML and DML to have separate implementations of `@trace` and `@param`. Prior to this, the `@trace` and `@param` macros directly produced code for the DML, and this led to issues with macro-expansion and parsing in the SML.

A more procedural description of how the `@gen` macro now works is as follows:

1. `macroexpand` the entire function body, with respect to the calling `__module__`. This expands any  (properly-scoped)`@trace` expressions to `Expr(:gentrace, ...)` expressions, and any (properly-scoped) `@param` expressions to `Expr(:genparam, ...)` expressions.
2. Desugar any tilde expressions `x ~ gen_fn()`, including those that may have been generated by macros, to `Expr(:gentrace, ...)` expressions. Scoping issues are avoided by desugaring to this custom expression that is not specific to any particular module.
3. Pass the macro-expanded and de-sugared function body on to `make_static_gen_function` or `make_dynamic_gen_function` accordingly.
4. For static gen functions, match any `:gentrace` expressions when adding address nodes to the DAG, and match `:genparam` expressions when adding param nodes to the DAG.
5. For dynamic gen functions, rewrite any `:gentrace` expression with its implementation `dynamic_trace_impl`, and rewrite any `:genparam` expression with its implementation `dynamic_param_impl`.

As in #278, we can now handle the following test case:

```julia
macro insert_normal_call(x)
    :(normal($(esc(x)), 0.0))
end

@gen (static) function foo()
    x ~ exponential(1)
    y ~ @insert_normal_call x
end
load_generated_macros()
trace = simulate(foo, ())
@test trace[:x] == trace[:y]
```

This works even if `Gen` is nested in another module.
```julia
module A import Gen end

macro insert_normal_call(x)
    :(A.Gen.normal($(esc(x)), 0.0))
end

A.Gen.@gen (static) function foo()
    x ~ exponential(1)
    y ~ @insert_normal_call x
end
A.Gen.load_generated_macros()
trace = A.Gen.simulate(foo, ())
@test trace[:x] == trace[:y]
```

 Thanks to @georgematheos for conversations + use cases that led to this approach.

### CODE REVIEW

**Review by georgematheos — APPROVED** [review]
@ztangent thanks for implementing this (and coming up with such an elegant design)!  All the code looks good to me.

**Comment by ztangent** [comment]
@marcoct @bzinberg should I wait for one of you to review this before merging?

**Comment by bzinberg** [comment]
@ztangent,

> @marcoct @bzinberg should I wait for one of you to review this before merging?

I'd say in general definitely best to get review before merging, all else being equal.  What time frame do you need for the review?

**Comment by bzinberg** [comment]
Ah, I see @georgematheos reviewed it already.  I'm definitely interested in checking it out as well if you're able to wait a week - are you?

**Comment by ztangent** [comment]
Sure! Just thought I'd check.

Xuan

**Comment by bzinberg** [comment]
That's very warranted :slightly_smiling_face: since we (to understate the situation) don't have official duties around reviewing PRs.

**Comment by georgematheos** [comment]
@ztangent also, let me write a few more test cases before we merge it; I'll try to do this later today.  I want to test macros that expand into `x ~ call(...)` syntax, `x = Gen.@trace(...)` syntax, and a few other cases like this.  I don't see why these should be an issue, but I was running into issues with some macros like this I was trying to write and I want to localize the problem.

**Comment by bzinberg** [comment]
Based on the PR description, this change is excellent and makes me feel like the create-your-own-DSL line of usage/research is on much firmer footing in terms of Julia integration (trying intercept macro calls before Julia saw them seemed brittle).  Awesome work!

Haven't reviewed the code yet but already have an ask :slightly_smiling_face: Could you copy your write-up of how Gen macro expansion and custom AST nodes work in the Gen doc site, perhaps in a "Developer Notes" section?

**Comment by ztangent** [comment]
@bzinberg docs added as requested!

**Comment by georgematheos** [comment]
I just made a PR into this `better-macro-support` branch with some test cases.  Most are passing, except the ones where I try to have a macro expand into `x ~ call(...)` syntax.  I'm going to try to investigate the issue and will report back if I can figure out what's causing this bug.

**Comment by georgematheos** [comment]
Okay, I figured out the issue I was running into and have resolved it in the test cases, so now everything is passing.  The issue is that if I write a macro along the lines of
```julia
macro draw_x()
  :( $(esc(:x)) ~ normal(0, 1) )
end
```
Macro hygiene will turn this into 
```julia
Expr(:call, Main.~, :x, :(normal(0, 1)))
```
The issue here is the `Main.~`, which does not get parsed by `desugar_tildes`.  The resolution is to change the macro so it reads
```julia
macro draw_x()
  Expr(:call, esc(:~), esc(:x), :(normal(0, 1)))
end
```

However, this is a bit annoying--as far as I'm aware, there is no way to inline an escaped tilde.  Ie. we can't use
```julia
:($(esc(:x)) esc(:~) normal(0, 1))
```
as the body of our macro, since while the `~` symbol is parsed as an infix notation function call, `esc(:~)` is not.

@ztangent do you think we should have the tilde desugaring support the use of `__module__.~` in addition to `~`, so the original version of this macro would work, and we don't have to go through these hoops?  Arguably, if we are using `~` syntax as a special symbol for Gen, we should parse it in the normal way even if it is bound to a certain function in the module.

**Comment by bzinberg** [comment]
@ztangent Thanks for adding those great docs!

@georgematheos, @ztangent, Oh wow, I did not realize the `~` captured by `@gen` this time was `Base.:~`, the bitwise-not operator!  I definitely think we should change it to instead capture a new `Gen.:~` which is exported by the `Gen` package (so existing code that writes `using Gen` and doesn't rely on `~` for bitwise-not will not need to change).  WDYT?

File: src/dsl/dsl.jl
Comment (bzinberg) [inline src/dsl/dsl.jl:178]: Why `esc(:Any)` rather than `esc(Any)`?  (I would expect `esc(:Any)` to not work because `:Any` is not a type)

File: src/dsl/static.jl
Comment (bzinberg) [inline src/dsl/static.jl:3]: yas

File: src/dsl/static.jl
Comment (bzinberg) [inline src/dsl/static.jl:216]: yas

File: src/dynamic/dynamic.jl
Comment (bzinberg) [inline src/dynamic/dynamic.jl:73]: > ```julia
> GlobalRef(@__MODULE__, :traceat)
> ```

:+1: :bug:

**Comment by georgematheos** [comment]
@bzinberg that sounds reasonable to me, and if implemented right it would probably resolve the issue I was encountering since it would bind `~` to the correct value in `__module__`.

However, it's worth noting that we are really not trying to use `~` as a function (like bitwise or), but instead as a piece of syntax.  So in a sense the question about what "namespace the `~` is from" seems strange to me, like asking "what namespace is the `.` syntax from"?  If we're committing to using `~` syntax, maybe we should just capture any `~` regardless of it's namespace, and tell users if they want "bitwise or" they have to use different syntax.  (A question here is: should we capture the syntax `MODULE.~`?  Or only `~` without a modulename?  If the latter, we have to deal with uglier syntax when writing macros...but in terms of simply "capturing the syntax" this seems to me like the most natural solution.)

Alternatively, we could try to limit the behavior of `desugar_tildes` so that it only desugars tildes included at the top-level expression of a line, so that if we have `~` used for bitwise or it works, eg. so
```julia
x ~ normal( 1 ~ fibonacci(3) , 2)
```
is desugared to
```julia
x = @trace(normal(Base.:~(1, fibonacci(3)), 2), :x)
```

File: test/runtests.jl
Comment (bzinberg) [inline test/runtests.jl]: Optional: i.m.o. it would be more readable to move this (tiny) macro definition to right before the tests that use it (i.e. have it appear in both files) rather than having them refer to the function definition here.  Ex. [rationale](https://evgenii.com/blog/code-duplication-in-unit-tests/)

File: src/dsl/dynamic.jl
Comment (bzinberg) [inline src/dsl/dynamic.jl:21]: For my own education: is `postwalk` an arbitrary choice, or is it the only kind of walk that would work here?

**Review by bzinberg — COMMENTED** [review]
Awesome!

**Comment by bzinberg** [comment]
@georgematheos 

> However, it's worth noting that we are really not trying to use `~` as a function (like bitwise or), but instead as a piece of syntax. So in a sense the question about what "namespace the `~` is from" seems strange to me, like asking "what namespace is the `.` syntax from"? If we're committing to using `~` syntax, maybe we should just capture any `~` regardless of it's namespace, and tell users if they want "bitwise or" they have to use different syntax.

This sounds right to me, and we can also expose a symbol like `Gen.bitwise_not = Base.:~` which can be used inside a `@gen` function body.  But is it actually possible to capture exactly all tildes?  I didn't realize macros could inspect the symbol used in a referentially-nontransparent way.

**Comment by georgematheos** [comment]
@bzinberg my understanding is that we are currently using a line like
```julia
MacroTools.@capture(e, lhs_Symbol ~ rhs_call)
```
which will find any time we have a `~` in infix notation between a symbol (eg. `x`) and a call (eg. `normal(0, 1)).
If we explicitly state a modulename, we would have to write
```julia
Main.:~(x, normal(0, 1))
```
since Julia cannot parse `Main.:~` in infix notation.  So the only time we can actually have `~` appear between the left hand side and the right hand side is then `~` is associated with the module at which a generative function is written (`__module__` in the macro).  So if all we want to support is this infix `~` notation, I think the current implementation handles this pretty much correctly--since if we are specifying a modulename, we have to use function call syntax, which we're not currently capturing.

If we wanted to support this function call syntax, we could add lines to the parsing like
```julia
MacroTools.@capture(e, modulename_.:~(lhs_Symbol, rhs_call))
```
The upside of doing this is it would make it marginally easier to write macros as I described above.  However, seems to me that this also moves us conceptually from "we support `x ~ call()` syntax" to "we overload the function `~` for syntactic sugar", which I'm not sure we want, as you're pointing out.

**Comment by bzinberg** [comment]
That might be right!  I'm not familiar enough with MacroTools at the moment to know how exactly the capture works.  I'm pretty sure though that the equivalence of `a ~ b` with `~(a, b)` is a hard-coded thing, specific to a certain set of symbols, as described [here](https://discourse.julialang.org/t/list-of-binary-infix-operators/32282).  IIRC, `~` will obey scoping rules for real reasons, but will be usable as an infix operator because of that specific hard-coded thing.

**Comment by georgematheos** [comment]
Right, to be more clear, I think `~(a, b)` and `a ~ b` are probably parsed the same way by MacroTools; but since we're only checking for `a_ ~ b_`, we won't pick up any `ModuleName.:~(a, b)` calls.

**Comment by ztangent** [comment]
Thanks for isolating the bug @georgematheos! As @georgematheos says, it looks like `a ~ b` and `~(a, b)` are parsed to the same `Expr`:

```julia
julia> dump(:(a ~ b))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ~
    2: Symbol a
    3: Symbol b

julia> dump(:(~(a, b)))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ~
    2: Symbol a
    3: Symbol b
```

On the other hand, a qualified reference to `~` parses differently
```julia
julia> dump(:(Gen.:~(a, b)))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Expr
      head: Symbol .
      args: Array{Any}((2,))
        1: Symbol Gen
        2: QuoteNode
          value: Symbol ~
    2: Symbol a
    3: Symbol b
```
which means that `MacroTools.@capture(e, lhs_ ~ rhs_)` will match the former two cases, but not the latter.

**Comment by ztangent** [comment]
My take on what we should do here: I agree with @georgematheos  that we should treat `a ~ b` as a new piece of Gen-specific syntax, rather than treating it as a function defined in `Gen` that needs to be qualified. I think as long as we document it clearly, this is a reasonable restriction that `@gen` functions can place on what kind of Julia syntax you can use. I say this especially because:

1. Our implementation of `desugar_tildes` doesn't actually match against the use of `~` as a unary operator, so users are actually still free to use it to mean "bitwise-not" (`Base.:~`), i.e., I could write something like, `{:x} ~ normal(~1, ~(-5))` and that should correctly desugar to `Expr(:gentrace, :(normal(~1, ~(-5))), :x)`.
2. While it is possible that users might want to define their own function `MyModule.:~(a, b)` which acts as a binary operator, this seems highly unlikely, and its a rare enough use case that I feel comfortable just documenting that we don't support using it as an infix operator. (Users can always use `MyModule.:~` in qualified non-infix form, because this is not captured by `desugar_tildes`)

@georgematheos regarding the difficulty of writing macros that introduce tilde expressions, I'm inclined to say that this is an advanced enough use case that we shouldn't be optimizing for user-friendliness here, at the expense of harder-to-maintain code. I'd go so far as to say that anyone who wants to write macros that introduce Gen-specific statements should understand Gen's parser enough that they know they can write a macro that introduces a `:gentrace` expression, instead of a tilde expression. In general, macros are fairly hard to write and get working, and I think we can expect advanced users (like ourselves) to do the debugging necessary to figure out what's needed to make things work. 

Tagging @alex-lew in case he thought about any of this when first introducing the tilde syntax.

**Comment by ztangent** [comment]
As evidence of point 1 above:

```julia
julia> expr = :({:x} ~ normal(~1, ~(-5)))
julia> Gen.desugar_tildes(expr)
:($(Expr(:gentrace, :(normal(~1, ~-5)), Some(:(:x)))))
```

**Comment by bzinberg** [comment]
> we should treat `a ~ b` as a new piece of Gen-specific syntax, rather than treating it as a function defined in `Gen` that needs to be qualified.

:+1: Yes, in fact it can't be just a function, because the syntax `x ~ normal(0, 1)` assigns to the local variable `x` so is not referentially transparent.  The part that is awkward i.m.o. (but probably ok) is that special treatment of `~` "really belongs" in the lexer, and we're working around that by saying that "`~` will always lex to the name `~` within some module [rather than to some special non-name lexeme], so we'll catch it the parse tree that way."

Edit: I think the part about lexing and `Module.:~` is wrong, as explained in https://github.com/probcomp/Gen.jl/pull/280#discussion_r456194349.

**Comment by bzinberg** [comment]
FWIW, `ShapesWorld.⊗` exists :slightly_smiling_face:

File: src/dsl/dsl.jl
Comment (ztangent) [reply src/dsl/dsl.jl:70]: Nope, prior to macro-expansion, `*` simply matches to an unqualified occurrence of `*`:
```julia
julia> dump(:({*}))
Expr
  head: Symbol braces
  args: Array{Any}((1,))
    1: Symbol *
```
If we want to match qualified occurrences of `*`, e.g., `A.B.Gen.*`, we have to add another branch in the if-then-else statement which does that specifically, and make sure that `A.B.Gen` is an appropriately qualified reference to the actual `Gen` module. I think we should avoid doing this and simply treat `{*}` as special syntax -- there's no use case in which we would want a `*` in the middle of two braces to refer to `Base.:*`.

File: src/dsl/dsl.jl
Comment (ztangent) [reply src/dsl/dsl.jl:115]: Yes, tilde expressions aren't processed at all until macro-expansion is completely finished. So you can introduce tilde-expressions using a macro, just like @georgematheos does in this comment: https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659008917

File: src/dsl/dsl.jl
Comment (ztangent) [reply src/dsl/dsl.jl:178]: The `esc` needs to be there, otherwise macro-expansion hygiene automatically adds a `Gen` as a qualifier. With `esc`:
```julia
julia> macroexpand(Main, :(@param theta))
:($(Expr(:genparam, :theta, :Any)))
```
Without `esc` (i.e., `macro param(sym::Symbol) = Expr(:genparam, esc(sym), :Any)`):
```julia
julia> macroexpand(Main, :(@param theta))
:($(Expr(:genparam, :theta, :(Gen.Any))))
```

Comment (ztangent) [reply src/dsl/dsl.jl:178]: Also, `:genparam` expressions are supposed to take only `Symbol`s as arguments --- it's the job of the of the SML and DML implementations to take those Symbols and appropriately resolve them to types. (At least in the current implementation of the SML parser. The DML just ignores type information for parameters.)

File: src/dsl/dynamic.jl
Comment (ztangent) [reply src/dsl/dynamic.jl:21]: You could also use `prewalk` here, since the expression-rewriting doesn't cause any infinite loops, but `postwalk` is recommended by default --- `postwalk` walks the AST bottom-up, and so it's easier to reason about what it's doing. I've only had to use `prewalk` when walking the AST top-down was specifically required, e.g. when doing `extract_quoted_expressions`.

File: src/dsl/dsl.jl
Comment (bzinberg) [reply src/dsl/dsl.jl:70]: Hm.  Weird that `~` and `*` seem to be treated differently.

Comment (ztangent) [reply src/dsl/dsl.jl:70]: They aren't! They're treated the same. How do you understand them as being treated differently?

Comment (bzinberg) [reply src/dsl/dsl.jl:70]: Ohh, after re-reading https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659008917 I now realize that the bug was in the macro that produced something that wasn't actually a plain `~`, i.e. that issue is independent of Gen.  Got it!

Comment (bzinberg) [reply src/dsl/dsl.jl:70]: Does that mean https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659657114 is wrong?  I.e. at what point during lexing, parsing, macro expansion, evaluation would `~` get turned into `Main.:~` if we didn't intervene?  Seems like it would have to be either evaluation or compilation time, i.e. after macro expansion.

Comment (ztangent) [reply src/dsl/dsl.jl:70]: Macro-expansion itself turns `~` into `Main.:~`:

```julia
julia> macro draw_x()
         :( $(esc(:x)) ~ normal(0, 1) )
       end
@draw_x (macro with 1 method)
julia> dump(macroexpand(Main, :(@draw_x())))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: GlobalRef
      mod: Module Main
      name: Symbol ~
    2: Symbol x
    3: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: GlobalRef
          mod: Module Main
          name: Symbol normal
        2: Int64 0
        3: Int64 1
```

You can see in the dumped Expression that `~` has become a `GlobalRef` with module `Main` and symbol `~`.

Comment (bzinberg) [reply src/dsl/dsl.jl:70]: It seems in that example that the resolution of `~` to `Main.:~` happened at macro definition time, not expansion time, right?  So macro expansion just substituted the already-resolved name.  I think in https://github.com/probcomp/Gen.jl/pull/280#discussion_r456194349 I had in mind an example like
```julia
macro draw_x()
  Expr(:~, esc(:x), normal(0, 1) )
end
```
but I don't know how to make an actually correct version of that example.

So maybe the full answer is: If the body of a `@gen` that contain tildes, those are captured directly, and the trick of treating specially `Module.:~` for every `Module` is not used.  If the `@gen` function body itself contains macros that expand to produce tildes, the code as currently written tries to actively correct places where the macro author used `~` in the macro definition (resolved to `Module.~` at macro definition time) instead of using the raw `:~` symbol.

If the above sounds right, I think we might want to drop the active correction and instead require macro authors to fix their macros?

[Edit: was previously off-topic, but is now possibly back on-topic!]  Thanks for the explanations so far!

cc @georgematheos

**Comment by georgematheos** [comment]
@ztangent that all makes sense to me!  (Especially since `~` is bitwise not, which is unary--for some reason I had thought it was bitwise or, which as a binary operator might have more overlap with our syntactic uses.)

My understanding is that this means the current state of the code is what we're going for (if Alex & Ben, and Marco if he reviews, are also comfortable with this).

File: src/dsl/dsl.jl
Comment (bzinberg) [reply src/dsl/dsl.jl:70]: Oh, this is exactly what @georgematheos was saying in https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659008917 :slowbro:

Comment (bzinberg) [reply src/dsl/dsl.jl:70]: I'm definitely in favor of dropping the preemptive intercepting of `Module.:~` (unless there are further reasons we need it beyond the above described).  We'd be breaking legitimate (albeit weird with poor name choices) uses of Julia semantics.  It could be helpful to check for possible `~`-related mistakes and issue a `@warning` if found.  But I think hypercorrecting potentially-correct-but-weird programs is worse than crashing.

Comment (bzinberg) [reply src/dsl/dsl.jl:70]: This is exactly what @ztangent said in https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659652523 :slowwwbrooo:

**Review by bzinberg — APPROVED** [review]
Looks good to me!  Really appreciated all your patient replies in https://github.com/probcomp/Gen.jl/pull/280#discussion_r455760791 and elsewhere.

@georgematheos, are you planning to merge in your extra unit tests once this branch enters master?

**Comment by ztangent** [comment]
Just merged your tests @georgematheos! @bzinberg before merging into master, I'm just going to document that we don't support user-definitions of tilde as an infix operator within `@gen` functions.

**Pull request merged**
